src/workerd/jsg/async-context.h

```diff
@@ -0,0 +1,178 @@
+  // storage context. So, for instance, when we start executing, the Root Frame's storage
+  // context is active. When a timeout elapses and a timer is going to fire, we enter the
+  // timer's Frame which makes that frame's storage context active. Once the timer
+  // callback has completed, we return back to the Root frame and storage context.
+  //
+  // All frames (except for the Root) are created within the scope of a parent, which by
+  // default is whichever frame is current when the new frame is created. When the new frame
+  // is created, it inherits a copy storage context of the parent.
+public:
+  class StorageKey: public kj::Refcounted {
+    // An opaque key that identifies an async-local storage cell within the frame.
+  public:
+    StorageKey() : hash(kj::hashCode(this)) {}
+
+    void reset() { dead = true; }
+    // The owner of the key should reset it when it goes away.
+    // The StorageKey is typically owned by an instance of AsyncLocalstorage (see
+    // the api/node/async-hooks.h). When the ALS instance is garbage collected, it
+    // must call reset to signal that this StorageKey is "dead" and can never be
+    // looked up again. Subsequent accesses to a frame will remove dead keys from
+    // the frame lazily. The lazy cleanup does mean that values may persist in
+    // memory a bit longer so if it proves to be problematic we can make the cleanup
+    // a bit more proactive.
+    //
+    // TODO(later): We should also evaluate the relatively unlikely case where an
+    // ALS is capturing a reference to itself and therefore can never be cleaned up.
+
+    bool isDead() const { return dead; }
+    inline uint hashCode() const { return hash; }
+    inline bool operator==(const StorageKey& other) const {
+      return hash == other.hash;
+    }
+
+  private:
+    uint hash;
+    bool dead = false;
+  };
+
+  struct StorageEntry {
+    kj::Own<StorageKey> key;
+    Value value;
+
+    inline StorageEntry clone(Lock& js) {
+      return {
+        .key = addRef(*key),
+        .value = value.addRef(js)
+      };
+    }
+  };
+
+  AsyncContextFrame(IsolateBase& isolate);
+  AsyncContextFrame(
+      Lock& js,
+      kj::Maybe<AsyncContextFrame&> maybeParent = nullptr,
+      kj::Maybe<StorageEntry> maybeStorageEntry = nullptr);
+
+  static AsyncContextFrame& current(Lock& js);
+  // Returns the reference to the AsyncContextFrame currently at the top of the stack.
+
+  static kj::Own<AsyncContextFrame> create(
+      Lock& js,
+      kj::Maybe<AsyncContextFrame&> maybeParent = nullptr,
+      kj::Maybe<StorageEntry> maybeStorageEntry = nullptr);
+  // Create a new AsyncContextFrame. If maybeParent is not specified, uses the current().
+  // If maybeStorageEntry is non-null, the associated storage cell in the new frame is
+  // set to the given value.
+
+  static v8::Local<v8::Function> attachContext(Lock& js, v8::Local<v8::Function> fn,
```

Comment:
For the version that takes a function, I think the name `wrap` actually made a lot more sense -- that's what it's doing, wrapping the function and returning a new function. I think this is very different from the `attachContext` that operates on promises so a different name makes sense.